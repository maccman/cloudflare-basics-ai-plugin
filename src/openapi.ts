import {
  OpenApiZodAny,
  generateSchema as generateBaseSchema,
} from '@anatine/zod-openapi'
import { RequestMethod, json } from 'cloudflare-basics'
import { OpenAPIObject, ParameterObject, PathsObject } from 'openapi3-ts/oas31'

export interface OpenApiPath {
  path: string
  method: RequestMethod
  operationId?: string
  tags?: string[]
  summary: string
  parameters?: ParameterObject
  requestBody: OpenApiZodAny
  responseBody: OpenApiZodAny
}

export interface OpenApiOptions {
  title: string
  description: string
  version: string
  paths: OpenApiPath[]
}

export function generateOpenApi(options: OpenApiOptions): OpenAPIObject {
  return {
    openapi: '3.0.2',
    info: {
      title: options.title,
      description: options.description,
      version: options.version,
    },
    paths: options.paths.reduce((paths, path) => {
      const { path: pathString, method } = path
      const pathObject = paths[pathString] || {}
      const methodObject = pathObject[method] || {}
      const requestBody = generateBaseSchema(path.requestBody)
      const responseBody = generateBaseSchema(path.responseBody)

      const operationId =
        path.operationId ?? `${method}_${pathString.replace(/\//g, '_')}')}`

      return {
        ...paths,
        [pathString]: {
          ...pathObject,
          [method]: {
            ...methodObject,
            operationId,
            ...(path.tags && { tags: path.tags }),
            summary: path.summary,
            requestBody: {
              content: {
                'application/json': {
                  schema: requestBody,
                },
              },
            },
            responses: {
              200: {
                description: 'Successful Response',
                content: {
                  'application/json': {
                    schema: responseBody,
                  },
                },
              },
            },
            ...(path.parameters && { parameters: path.parameters }),
          },
        },
      }
    }, {} as PathsObject),
  }
}

export function OpenApiRoute(options: OpenApiOptions) {
  return async () => {
    return json(generateOpenApi(options))
  }
}
