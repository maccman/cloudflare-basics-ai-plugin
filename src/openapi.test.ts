import { describe, expect, it } from 'vitest'
import { generateOpenApi } from './openapi'
import { z } from 'zod'

describe('generateOpenApi', () => {
  it('should generate an OpenAPI spec', () => {
    const openApi = generateOpenApi({
      title: 'Reflect Notes',
      description:
        'A plugin that allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.',
      version: '1.0.0',
      paths: [
        {
          path: '/notes',
          method: 'post',
          summary: 'Create a note',
          requestSchema: z.object({
            text: z.string(),
          }),
          responseSchema: z.object({
            id: z.string(),
            text: z.string(),
          }),
        },
      ],
    })

    expect(openApi).toMatchSnapshot()
  })
})
