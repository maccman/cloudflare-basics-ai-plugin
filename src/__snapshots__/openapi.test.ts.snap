// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateOpenApi > should generate an OpenAPI spec 1`] = `
{
  "info": {
    "description": "A plugin that allows the user to save notes. For example, saving a summary of their ChatGPT conversation history.",
    "title": "Reflect Notes",
    "version": "1.0.0",
  },
  "openapi": "3.0.2",
  "paths": {
    "/notes": {
      "post": {
        "operationId": "post__notes')}",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "text": {
                    "type": "string",
                  },
                },
                "required": [
                  "text",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "type": "string",
                    },
                    "text": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "id",
                    "text",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Successful Response",
          },
        },
        "summary": "Create a note",
      },
    },
  },
}
`;
